Can't create a generic service with impl_web?

Hello, I'm trying to create a service that has a generic parameter and I'm getting an error. Here is some example code:

```
#[macro_use] extern crate tower_web;
use std::error::Error;
use std::fmt::Display;

struct HelloWorld<S>(pub S);

impl_web! {
    impl<S: Display> HelloWorld<S> {
        #[get("/")]
        fn hello(&self) -> Result<String, Box<Error>> {
            let r = self.0.to_string();
            Ok(r)
        }
    }
}

fn main() {
    println!("Hello, world!");
}
```

and here is the error when I go to compile it:

```
~/c/test-tower-project> cargo check
    Checking test-tower-project v0.1.0 (/home/paul/code/test-tower-project)                                                                                                                   
error[E0412]: cannot find type `S` in this scope                                                                                                                                              
  --> src/main.rs:7:1                                                                                                                                                                         
   |                                                                                                                                                                                          
7  | / impl_web! {                                                                                                                                                                            
8  | |     impl<S: Display> HelloWorld<S> {                                                                                                                                                   
9  | |         #[get("/")]                                                                                                                                                                    
10 | |         fn hello(&self) -> Result<String, Box<Error>> {                                                                                                                                
...  |                                                                                                                                                                                        
14 | |     }                                                                                                                                                                                  
15 | | }                                                                                                                                                                                      
   | |_^ did you mean `U`?                                                                                                                                                                    
   |                                                                                                                                                                                          
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)                                              
                                                                                                                                                                                              
error[E0599]: no method named `hello` found for type `HelloWorld<S>` in the current scope                                                                                                     
  --> src/main.rs:7:1                                                                                                                                                                         
   |                                                                                                                                                                                          
5  |   struct HelloWorld<S>(pub S);                                                                                                                                                           
   |   ---------------------------- method `hello` not found for this                                                                                                                         
6  |                                                                                                                                                                                          
7  | / impl_web! {                                                                                                                                                                            
8  | |     impl<S: Display> HelloWorld<S> {                                                                                                                                                   
9  | |         #[get("/")]                                                                                                                                                                    
10 | |         fn hello(&self) -> Result<String, Box<Error>> {                                                                                                                                
...  |                                                                                                                                                                                        
14 | |     }                                                                                                                                                                                  
15 | | }                                                                                                                                                                                      
   | |_^                                                                                                                                                                                      
   |                                                                                                                                                                                          
   = note: the method `hello` exists but the following trait bounds were not satisfied:                                                                                                       
           `S : std::fmt::Display`                                                                                                                                                            
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)                                              
                                                                                                                                                                                              
error: aborting due to 2 previous errors                                                                                                                                                      
                                                                                                                                                                                              
Some errors occurred: E0412, E0599.                                                                                                                                                           
For more information about an error, try `rustc --explain E0412`.                                                                                                                             
error: Could not compile `test-tower-project`.                                                                                                                                                

To learn more, run the command again with --verbose.
```

Is there a way around this? I assume it's a problem with the proc-macro but I'm really not sure what to do about it, other than to have my service use boxed trait objects instead of monomorphization.

--- a/tests/resource.rs
+++ b/tests/resource.rs
@@ -1,7 +1,11 @@
+#![recursion_limit="256"]
+
 extern crate futures;
 extern crate http;
 #[macro_use]
 extern crate tower_web;
+
+use std::fmt::Display;
 
 #[macro_use]
 mod support;
@@ -9,6 +13,18 @@ use support::*;
 
 #[derive(Clone, Debug)]
 struct TestResource;
+
+#[derive(Clone, Debug)]
+struct HelloWorld<S>(S);
+
+#[derive(Clone, Debug)]
+struct WhereHelloWorld<S>(S);
+
+#[derive(Clone, Debug)]
+struct EmptyGeneric<S>(S);
+
+#[derive(Clone, Debug)]
+struct EmptyWhere<S>(S);
 
 impl_web! {
     impl TestResource {
@@ -19,6 +35,32 @@ impl_web! {
             Ok("impl_future").into_future()
         }
     }
+
+    impl<S: Display> HelloWorld<S> {
+        #[get("/")]
+        fn hello(&self) -> Result<String, ()> {
+            let r = self.0.to_string();
+            Ok(r)
+        }
+    }
+
+    impl<S> WhereHelloWorld<S>
+    where
+        S: Display,
+    {
+        #[get("/")]
+        fn hello(&self) -> Result<String, ()> {
+            let r = self.0.to_string();
+            Ok(r)
+        }
+    }
+
+    // Just make sure this compiles
+    impl<S> EmptyGeneric<S> {
+    }
+
+    impl<S> EmptyWhere<S> where S: Display {
+    }
 }
 
 #[test]
@@ -28,4 +70,19 @@ fn impl_future() {
     let response = web.call_unwrap(get!("/impl_future"));
     assert_ok!(response);
     assert_body!(response, "impl_future");
+}
+
+#[test]
+fn generic() {
+    let mut web = service(HelloWorld("hello"));
+
+    let response = web.call_unwrap(get!("/"));
+    assert_ok!(response);
+    assert_body!(response, "hello");
+
+    let mut web = service(WhereHelloWorld("hello"));
+
+    let response = web.call_unwrap(get!("/"));
+    assert_ok!(response);
+    assert_body!(response, "hello");
 }

--- a/tower-web-macros/src/resource/parse.rs
+++ b/tower-web-macros/src/resource/parse.rs
@@ -65,8 +65,20 @@ impl syn::fold::Fold for ImplWeb {
             "trait impls must not be in impl_web! block"
         );
 
+        // The resource may only have type generics (for now at least).
+        for param in &item.generics.params {
+            use syn::GenericParam::Type;
+
+            match *param {
+                Type(_) => {}
+                ref actual => {
+                    panic!("Resources may only have generic type parameters. Actual = {:?}", actual);
+                }
+            }
+        }
+
         let index = self.resources.len();
-        self.resources.push(Resource::new(index, item.self_ty.clone()));
+        self.resources.push(Resource::new(index, &item));
 
         syn::fold::fold_item_impl(self, item)
     }

--- a/tower-web-macros/src/resource/resource.rs
+++ b/tower-web-macros/src/resource/resource.rs
@@ -12,6 +12,9 @@ pub(crate) struct Resource {
     /// The type implementing `Resource`
     pub self_ty: Box<syn::Type>,
 
+    /// Resource generics
+    generics: syn::Generics,
+
     /// The route handlers implemented by `Resource`
     pub routes: Vec<Route>,
 
@@ -28,10 +31,14 @@ struct Destination {
 }
 
 impl Resource {
-    pub fn new(index: usize, self_ty: Box<syn::Type>) -> Resource {
+    pub fn new(index: usize, item: &syn::ItemImpl) -> Resource {
+        let self_ty = item.self_ty.clone();
+        let generics = item.generics.clone();
+
         Resource {
             index,
             self_ty,
+            generics,
             routes: vec![],
             catches: vec![],
             destinations: vec![],
@@ -51,6 +58,8 @@ impl Resource {
 
         let dummy_const = self.dummy_const();
         let ty = &self.self_ty;
+        let generics = self.resource_generics();
+        let where_predicates = self.where_predicates();
 
         quote! {
             #[allow(warnings)]
@@ -59,10 +68,13 @@ impl Resource {
 
                 #resource_impl
 
-                impl<U> __tw::util::Chain<U> for #ty {
-                    type Output = (Self, U);
+                impl<__U, #generics> __tw::util::Chain<__U> for #ty
+                where
+                    #where_predicates
+                {
+                    type Output = (Self, __U);
 
-                    fn chain(self, other: U) -> Self::Output {
+                    fn chain(self, other: __U) -> Self::Output {
                         (self, other)
                     }
                 }
@@ -72,6 +84,9 @@ impl Resource {
 
     fn gen_impl(&self) -> TokenStream {
         let ty = &self.self_ty;
+
+        let generics = self.resource_generics();
+        let where_predicates = self.where_predicates();
 
         // The destination type is `Either{N}` where `N` is the number of routes
         // or ::MAX_VARIANTS, which ever is smaller. In the latter case, some
@@ -120,18 +135,18 @@ impl Resource {
 
             #async_helper_macro
 
-            pub struct GeneratedResource<S, B>
+            pub struct GeneratedResource<S, B, T>
             where S: __tw::response::Serializer,
                   B: __tw::util::BufStream,
             {
-                inner: ::std::sync::Arc<Inner<S>>,
+                inner: ::std::sync::Arc<Inner<S, T>>,
                 _p: ::std::marker::PhantomData<B>,
             }
 
-            struct Inner<S>
+            struct Inner<S, T>
             where S: __tw::response::Serializer,
             {
-                handler: #ty,
+                handler: T,
                 callsites: CallSites,
                 content_types: ContentTypes<S>,
                 serializer: S,
@@ -140,11 +155,11 @@ impl Resource {
             #callsites_def
             #content_types_def
 
-            impl<S, B> GeneratedResource<S, B>
+            impl<S, B, T> GeneratedResource<S, B, T>
             where S: __tw::response::Serializer,
                   B: __tw::util::BufStream,
             {
-                fn new(handler: #ty, serializer: S) -> Self {
+                fn new(handler: T, serializer: S) -> Self {
                     let callsites = CallSites::new::<B>();
                     let content_types = ContentTypes::new(&serializer);
 
@@ -164,7 +179,7 @@ impl Resource {
                 }
             }
 
-            impl<S, B> Clone for GeneratedResource<S, B>
+            impl<S, B, T> Clone for GeneratedResource<S, B, T>
             where S: __tw::response::Serializer,
                   B: __tw::util::BufStream,
             {
@@ -177,12 +192,13 @@ impl Resource {
                 }
             }
 
-            impl<S, B> __tw::routing::IntoResource<S, B> for #ty
-            where S: __tw::response::Serializer,
-                  B: __tw::util::BufStream,
+            impl<__S, __B, #generics> __tw::routing::IntoResource<__S, __B> for #ty
+            where __S: __tw::response::Serializer,
+                  __B: __tw::util::BufStream,
+                  #where_predicates
             {
                 type Destination = #destination_ty;
-                type Resource = GeneratedResource<S, B>;
+                type Resource = GeneratedResource<__S, __B, #ty>;
 
                 fn routes(&self) -> __tw::routing::RouteSet<Self::Destination> {
                     __tw::routing::Builder::new()
@@ -190,20 +206,21 @@ impl Resource {
                     .build()
                 }
 
-                fn into_resource(self, serializer: S) -> Self::Resource {
+                fn into_resource(self, serializer: __S) -> Self::Resource {
                     GeneratedResource::new(self, serializer)
                 }
             }
 
-            impl<S, B> __tw::routing::Resource for GeneratedResource<S, B>
-            where S: __tw::response::Serializer,
-                  B: __tw::util::BufStream,
+            impl<__S, __B, #generics> __tw::routing::Resource for GeneratedResource<__S, __B, #ty>
+            where __S: __tw::response::Serializer,
+                  __B: __tw::util::BufStream,
+                  #where_predicates
             {
                 // The destination token is used to identify which action to
                 // call
                 type Destination = #destination_ty;
 
-                type RequestBody = B;
+                type RequestBody = __B;
 
                 // The response body's chunk type.
                 type Buf = <Self::Body as __tw::util::BufStream>::Item;
@@ -212,7 +229,7 @@ impl Resource {
                 type Body = <Self::Future as __tw::routing::ResourceFuture>::Body;
 
                 // Future representing processing the request.
-                type Future = ResponseFuture<S, B>;
+                type Future = ResponseFuture<__S, __B, #ty>;
 
                 fn dispatch(
                     &mut self,
@@ -228,13 +245,12 @@ impl Resource {
 
             #catch_impl
 
-            pub struct ResponseFuture<S, B>
+            pub struct ResponseFuture<S, B, T>
             where S: __tw::response::Serializer,
                   B: __tw::util::BufStream,
             {
                 state: State<B>,
-                inner: ::std::sync::Arc<Inner<S>>,
-                _p: ::std::marker::PhantomData<B>,
+                inner: ::std::sync::Arc<Inner<S, T>>,
             }
 
             // Tracks the resource's response state. At a high level, the steps
@@ -257,9 +273,10 @@ impl Resource {
                 Invalid(::std::marker::PhantomData<B>),
             }
 
-            impl<S, B> __tw::routing::ResourceFuture for ResponseFuture<S, B>
-            where S: __tw::response::Serializer,
-                  B: __tw::util::BufStream,
+            impl<__S, __B, #generics> __tw::routing::ResourceFuture for ResponseFuture<__S, __B, #ty>
+            where __S: __tw::response::Serializer,
+                  __B: __tw::util::BufStream,
+                  #where_predicates
             {
                 type Body = ResponseBody;
 
@@ -411,20 +428,23 @@ impl Resource {
     /// Generate code for a resource with no routes
     fn gen_empty_impl(&self) -> TokenStream {
         let ty = &self.self_ty;
+        let generics = self.resource_generics();
+        let where_predicates = self.where_predicates();
 
         quote! {
-            impl<S, B> __tw::routing::IntoResource<S, B> for #ty
-            where S: __tw::response::Serializer,
-                  B: __tw::util::BufStream,
+            impl<__S, __B, #generics> __tw::routing::IntoResource<__S, __B> for #ty
+            where __S: __tw::response::Serializer,
+                  __B: __tw::util::BufStream,
+                  #where_predicates
             {
                 type Destination = ();
-                type Resource = __tw::routing::Unit<B>;
+                type Resource = __tw::routing::Unit<__B>;
 
                 fn routes(&self) -> __tw::routing::RouteSet<Self::Destination> {
                     __tw::routing::RouteSet::new()
                 }
 
-                fn into_resource(self, serializer: S) -> Self::Resource {
+                fn into_resource(self, serializer: __S) -> Self::Resource {
                     __tw::routing::Unit::new()
                 }
             }
@@ -439,6 +459,27 @@ impl Resource {
                 Some(&segments[len-1].ident)
             }
             _ => None,
+        }
+    }
+
+    fn resource_generics(&self) -> TokenStream {
+        let generics = self.generics.params.iter();
+
+        quote! {
+            #(#generics),*
+        }
+    }
+
+    fn where_predicates(&self) -> TokenStream {
+        if let Some(ref clause) = self.generics.where_clause {
+            let predicates = clause.predicates.iter();
+
+            quote! {
+                #(#predicates),*
+            }
+        } else {
+            quote! {
+            }
         }
     }
 
@@ -606,7 +647,6 @@ impl Resource {
             ResponseFuture {
                 state,
                 inner,
-                _p: ::std::marker::PhantomData,
             }
         }
     }

--- a/tower-web-macros/src/resource/ty_tree.rs
+++ b/tower-web-macros/src/resource/ty_tree.rs
@@ -67,11 +67,11 @@ impl<'a> TyTree<'a, Arg> {
                         &callsites.#index.0);
 
                     if callsites.#index.1 {
-                        <#ty as __tw::extract::Extract<B>>::extract_body(
+                        <#ty as __tw::extract::Extract<__B>>::extract_body(
                             &context,
                             body.take().unwrap())
                     } else {
-                        <#ty as __tw::extract::Extract<B>>::extract(&context)
+                        <#ty as __tw::extract::Extract<__B>>::extract(&context)
                     }
                 }}
             },

